sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java

```diff
@@ -380,33 +403,47 @@ private HttpPipeline setupPipeline(HttpClient httpClient) {
-        return Mono.fromFuture(publicClientApplication.acquireToken(
+        return Mono.fromFuture(() -> getPublicClientApplication(false).acquireToken(
             UserNamePasswordParameters.builder(new HashSet<>(request.getScopes()), username, password.toCharArray())
                 .build()))
-            .map(ar -> new MsalToken(ar, options));
+            .onErrorMap(t -> new ClientAuthenticationException("Failed to acquire token with username and password",
+                null, t)).map(ar -> new MsalToken(ar, options));
     }
 
     /**
      * Asynchronously acquire a token from the currently logged in client.
      *
      * @param request the details of the token request
+     * @param account the account used to login to acquire the last token
      * @return a Publisher that emits an AccessToken
      */
-    public Mono<MsalToken> authenticateWithUserRefreshToken(TokenRequestContext request, MsalToken msalToken) {
-        SilentParameters parameters;
-        if (msalToken.getAccount() != null) {
-            parameters = SilentParameters.builder(new HashSet<>(request.getScopes()), msalToken.getAccount()).build();
-        } else {
-            parameters = SilentParameters.builder(new HashSet<>(request.getScopes())).build();
-        }
-        return Mono.defer(() -> {
+    public Mono<MsalToken> authenticateWithMsalAccount(TokenRequestContext request, IAccount account) {
+        return Mono.defer(() -> Mono.fromFuture(() -> {
+            SilentParameters.SilentParametersBuilder parametersBuilder = SilentParameters.builder(
+                new HashSet<>(request.getScopes()));
+            if (account != null) {
+                parametersBuilder = parametersBuilder.account(account);
+            }
             try {
-                return Mono.fromFuture(publicClientApplication.acquireTokenSilently(parameters))
-                        .map(ar -> new MsalToken(ar, options));
+                return getPublicClientApplication(false)
+                    .acquireTokenSilently(parametersBuilder.build());
             } catch (MalformedURLException e) {
-                return Mono.error(e);
+                throw logger.logExceptionAsError(Exceptions.propagate(e));
             }
-        });
+        }).map(ar -> new MsalToken(ar, options))
+        .filter(t -> !t.isExpired())
+        .switchIfEmpty(Mono.fromFuture(() -> {
+            SilentParameters.SilentParametersBuilder forceParametersBuilder = SilentParameters.builder(
+                new HashSet<>(request.getScopes())).forceRefresh(true);
+            if (account != null) {
+                forceParametersBuilder = forceParametersBuilder.account(account);
+            }
+            try {
+                return getPublicClientApplication(false).acquireTokenSilently(forceParametersBuilder.build());
+            } catch (MalformedURLException e) {
+                throw logger.logExceptionAsError(Exceptions.propagate(e));
```

Comment:
Since you're using futures, it may be easier to follow using:
```java
var completableFuture = new CompletableFuture<MsalToken>();
completableFuture.completeExceptionally(logger.logExceptionAsError(Exceptions.propagate(e));
return completableFuture;
```

Reply:
Thanks Connie for the tip but in this code the exception is thrown from creating the future, not from executing it.